### DISCUSSION TITLE
Custom middleware type conflicts with Jest

### DISCUSSION THREAD

**Discussion by dhmk083** [pmndrs/zustand#1876] [Bug report]
Maybe that's actually a Jest issue, not sure.

I used [an example of custom middleware that changes store type](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#middleware-that-changes-the-store-type) from docs in a Jest test. At first, it worked fine. But if I add `import { immer } from 'zustand/middleware/immer'`, then the test fails with an error: 

```
Overload 1 of 3, '(initializer: StateCreator<{ bears: number; }, [], ["zustand/immer", unknown][]>): UseBoundStore<StoreApi<{ bears: number; }>>', gave the following error.
        Argument of type 'StateCreator<{ bears: number; }, [], [["foo", string]]>' is not assignable to parameter of type 'StateCreator<{ bears: number; }, [], ["zustand/immer", unknown][]>'.
          Type 'StateCreator<{ bears: number; }, [], [["foo", string]]>' is not assignable to type '{ $$storeMutators?: ["zustand/immer", unknown][] | undefined; }'.
            Types of property '$$storeMutators' are incompatible.
              Type '[["foo", string]] | undefined' is not assignable to type '["zustand/immer", unknown][] | undefined'.
                Type '[["foo", string]]' is not assignable to type '["zustand/immer", unknown][]'.
                  Type '["foo", string]' is not assignable to type '["zustand/immer", unknown]'.
                    Type at position 0 in source is not compatible with type at position 0 in target.
                      Type '"foo"' is not assignable to type '"zustand/immer"'.
      Overload 2 of 3, '(store: StoreApi<unknown>): UseBoundStore<StoreApi<unknown>>', gave the following error.
        Argument of type 'StateCreator<{ bears: number; }, [], [["foo", string]]>' is not assignable to parameter of type 'StoreApi<unknown>'.
```

This error appears only when I run tests, vscode intellisense gives no error, also I can compile it with tsc without the error.

Here is a [demo](https://github.com/dhmk083/zustand-middleware-typescript). Run `npm test` to see the error, then comment immer import line and try again.

**Comment by dbritto-dev** [comment]
@dhmk083 thanks I'll look into that. Would you mid looking this thread ( https://github.com/pmndrs/zustand/discussions/1796#discussioncomment-5826280 ) for now?

**Comment by dhmk083** [comment]
I can't use workaround from that thread, because it's a middleware, I can't control how it could be combined with other middlewares.

By the way, I tried other middlewares like `devtools` and `persist`, and they caused the error too.

**Reply by dbritto-dev** [reply]
@dhmk083 yeah, you need to be careful while using middlewares and types I'll create an example for you

**Comment by dhmk083** [comment]
Good news, I fixed that error.

```ts
// fixed module name
declare module "zustand/vanilla" {
  interface StoreMutators<S, A> {
    foo: Write<Cast<S, object>, { foo: A }>;
  }
}
```

I changed module name from `zustand` to `zustand/vanilla`, since in source code it is declared as `../vanilla`. I'm happy now :)

**Reply by dbritto-dev** [reply]
@dhmk083 nice, keep going :D
